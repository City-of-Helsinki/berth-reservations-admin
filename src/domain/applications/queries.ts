import { gql } from 'apollo-boost';

export const BERTH_APPLICATIONS_QUERY = gql`
  query BERTH_APPLICATIONS {
    berthApplications {
      edges {
        node {
          id
          customer {
            id
          }
          berthSwitch {
            id
            berthNumber
            harbor
            harborName
            pier
          }
          createdAt
          municipality
          boatType
          boatRegistrationNumber
          boatWidth
          boatLength
          boatDraught
          boatWeight
          boatName
          boatModel
          accessibilityRequired
          status
          lease {
            id
            berth {
              number
              pier {
                properties {
                  identifier
                  harbor {
                    id
                    properties {
                      name
                    }
                  }
                }
              }
            }
          }
          harborChoices {
            harbor
            priority
            harborName
          }
        }
      }
    }
    boatTypes {
      id
      name
    }
  }
`;
