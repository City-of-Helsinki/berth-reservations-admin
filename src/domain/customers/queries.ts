import { gql } from 'apollo-boost';

export const CUSTOMER_QUERY = gql`
  query CUSTOMERS {
    profiles(serviceType: BERTH) {
      edges {
        node {
          id
          firstName
          lastName
          nickname
          primaryAddress {
            address
            city
            postalCode
          }
          primaryPhone {
            phone
          }
          primaryEmail {
            email
          }
          serviceConnections {
            edges {
              node {
                id
                service {
                  id
                  type
                }
              }
            }
          }
          contactMethod
          image
        }
      }
    }
  }
`;
