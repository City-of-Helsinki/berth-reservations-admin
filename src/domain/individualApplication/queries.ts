import { gql } from 'apollo-boost';

export const FILTERED_CUSTOMERS_QUERY = gql`
  query FILTERED_CUSTOMERS($firstName: String, $lastName: String) {
    profiles(
      serviceType: BERTH
      firstName: $firstName
      lastName: $lastName
      orderBy: "lastName"
    ) {
      edges {
        node {
          id
          firstName
          lastName
          primaryAddress {
            address
            postalCode
            city
          }
          organization {
            businessId
          }
          berthLeases {
            edges {
              node {
                berth {
                  number
                  pier {
                    properties {
                      identifier
                      harbor {
                        id
                        properties {
                          name
                        }
                      }
                    }
                  }
                }
              }
            }
          }
          berthApplications {
            edges {
              node {
                berthSwitch {
                  harborName
                }
              }
            }
          }
        }
      }
    }
  }
`;

export const INDIVIDUAL_APPLICATION_QUERY = gql`
  query INDIVIDUAL_APPLICATION($id: ID!) {
    berthApplication(id: $id) {
      id
      firstName
      lastName
      address
      municipality
      zipCode
      phoneNumber
      email

      customer {
        id
        organization {
          businessId
        }
      }
      berthSwitch {
        berthNumber
        harbor
        harborName
        id
        pier
        reason {
          title
        }
      }
      createdAt
      municipality
      boatType
      boatRegistrationNumber
      boatWidth
      boatLength
      boatDraught
      boatWeight
      boatName
      boatModel
      accessibilityRequired
      status
      lease {
        id
        berth {
          berthType {
            depth
            length
            mooringType
            width
          }
          comment
          isAccessible
          number
          pier {
            properties {
              identifier
              electricity
              gate
              lighting
              mooring
              wasteCollection
              water
              harbor {
                id
                properties {
                  name
                }
              }
            }
          }
        }
      }
      harborChoices {
        harbor
        priority
        harborName
      }
    }
    boatTypes {
      id
      name
    }
  }
`;
