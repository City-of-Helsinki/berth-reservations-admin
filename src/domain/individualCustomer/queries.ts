import { gql } from 'apollo-boost';

export const INDIVIDUAL_CUSTOMER_QUERY = gql`
  query INDIVIDUAL_CUSTOMER($id: ID!) {
    profile(id: $id, serviceType: BERTH) {
      id
      firstName
      lastName
      primaryEmail {
        email
      }
      primaryPhone {
        phone
      }
      primaryAddress {
        address
        postalCode
        city
      }
      invoicingType
      comment
      boats {
        edges {
          node {
            id
            boatType {
              id
              name
            }
            width
            length
            draught
            weight
            name
            model
            registrationNumber
            propulsion
            hullMaterial
          }
        }
      }
      berthLeases {
        edges {
          node {
            id
            status
            startDate
            endDate
            berth {
              number
              pier {
                properties {
                  identifier
                  harbor {
                    id
                    properties {
                      name
                    }
                  }
                }
              }
            }
          }
        }
      }
      berthApplications {
        edges {
          node {
            id
            berthSwitch {
              berthNumber
              harbor
              harborName
              id
              pier
              reason {
                title
              }
            }
            createdAt
            status
            lease {
              id
              berth {
                number
                pier {
                  properties {
                    identifier
                    harbor {
                      id
                      properties {
                        name
                      }
                    }
                  }
                }
              }
            }
            boatType
            boatRegistrationNumber
            boatWidth
            boatLength
            boatDraught
            boatWeight
            boatName
            boatModel
            harborChoices {
              harbor
              priority
              harborName
            }
            accessibilityRequired
          }
        }
      }
    }
    boatTypes {
      id
      name
    }
  }
`;
