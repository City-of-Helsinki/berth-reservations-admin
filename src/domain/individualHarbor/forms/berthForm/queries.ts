import { gql } from 'apollo-boost';

export const INDIVIDUAL_BERTH_QUERY = gql`
  query INDIVIDUAL_BERTH($id: ID!) {
    berth(id: $id) {
      number
      comment
      isActive
      pier {
        id
        properties {
          identifier
        }
      }
      mooringType
      width
      length
      depth
    }
  }
`;
