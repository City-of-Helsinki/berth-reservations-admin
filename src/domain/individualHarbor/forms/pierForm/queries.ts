import { gql } from 'apollo-boost';

export const BOAT_TYPES_QUERY = gql`
  query BOAT_TYPES {
    boatTypes {
      id
      name
    }
  }
`;
