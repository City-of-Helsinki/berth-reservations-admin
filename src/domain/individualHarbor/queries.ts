import { gql } from 'apollo-boost';

export const INDIVIDUAL_HARBOR_QUERY = gql`
  query INDIVIDUAL_HARBOR($id: ID!) {
    harbor(id: $id) {
      id
      properties {
        name
        numberOfPlaces
        numberOfFreePlaces
        streetAddress
        zipCode
        municipality
        wwwUrl
        imageFile
        servicemapId
        maxWidth
        piers {
          edges {
            node {
              properties {
                identifier
                electricity
                wasteCollection
                water
                lighting
                gate
                suitableBoatTypes {
                  name
                }
                berths {
                  edges {
                    node {
                      number
                      width
                      length
                      depth
                      mooringType
                      comment
                      leases {
                        edges {
                          node {
                            customer {
                              id
                              firstName
                              lastName
                            }
                            status
                            startDate
                            endDate
                            isActive
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
`;
