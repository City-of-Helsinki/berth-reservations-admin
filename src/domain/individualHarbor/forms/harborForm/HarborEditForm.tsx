import React, { FunctionComponent } from 'react';
import { Button, TextInput } from 'hds-react/lib';
import { useTranslation } from 'react-i18next';
import { Field, Form, Formik } from 'formik';
import * as Yup from 'yup';
import { TFunction } from 'i18next';
import { useMutation, useQuery } from '@apollo/react-hooks';

import styles from './harborEditForm.module.scss';
import Section from '../../../../common/section/Section';
import FileUpload from '../../../../common/fileUpload/FileUpload';
import Text from '../../../../common/text/Text';
import { INDIVIDUAL_HARBOR_QUERY } from '../../queries';
import LoadingSpinner from '../../../../common/spinner/LoadingSpinner';
import { INDIVIDUAL_HARBOR } from '../../__generated__/INDIVIDUAL_HARBOR';
import { Harbor } from './types';
import { getHarbor } from './utils/utils';

export interface HarborEditFormProps {
  onCancel: () => void;
  onSubmit: () => void;
  harborId: string;
}

const getValidationSchema = (t: TFunction) =>
  Yup.object<Harbor>().shape({
    name: Yup.string().required(t('forms.common.errors.required')),
    streetAddress: Yup.string().required(t('forms.common.errors.required')),
    zipCode: Yup.string().required(t('forms.common.errors.required')),
    municipality: Yup.string().required(t('forms.common.errors.required')),
    wwwUrl: Yup.string().required(t('forms.common.errors.required')),
    imageFile: Yup.object().required(t('forms.common.errors.required')),
  });

const HarborEditForm: FunctionComponent<HarborEditFormProps> = ({ onCancel, onSubmit, harborId }) => {
  const { loading, error, data } = useQuery<INDIVIDUAL_HARBOR>(INDIVIDUAL_HARBOR_QUERY, {
    variables: { id: harborId },
  });

  // TODO: Create update mutation

  const { t } = useTranslation();
  const validationSchema = getValidationSchema(t);

  if (loading) return <LoadingSpinner isLoading={loading} />;
  if (error) return <div>{t('forms.common.error')}</div>;

  const initial: Harbor = getHarbor(data) ?? {
    name: '',
    streetAddress: '',
    zipCode: '',
    municipality: 'Helsinki',
    wwwUrl: '',
    imageFile: undefined,
  };

  return (
    <Formik initialValues={initial} onSubmit={onSubmit} validationSchema={validationSchema}>
      {({ isSubmitting, errors, setFieldValue }) => (
        <Form>
          <Text as="h4" color="brand">
            {t('forms.harbor.title').toUpperCase()}
          </Text>

          <Section>
            <Field
              required={true}
              as={TextInput}
              id="name"
              name="name"
              labelText={t('forms.harbor.name')}
              invalid={!!errors.name}
              invalidText={errors.name}
            />
          </Section>

          <Section>
            <Field
              required={true}
              as={TextInput}
              id="streetAddress"
              name="streetAddress"
              labelText={t('forms.harbor.streetAddress')}
              invalid={!!errors.streetAddress}
              invalidText={errors.streetAddress}
            />
            <Field
              required={true}
              as={TextInput}
              id="zipCode"
              name="zipCode"
              labelText={t('forms.harbor.zipCode')}
              invalid={!!errors.zipCode}
              invalidText={errors.zipCode}
            />
            <Field
              required={true}
              as={TextInput}
              id="municipality"
              name="municipality"
              labelText={t('forms.harbor.municipality')}
              invalid={!!errors.municipality}
              invalidText={errors.municipality}
            />
          </Section>

          <Section>
            <Field
              required={true}
              as={TextInput}
              id="wwwUrl"
              name="wwwUrl"
              labelText={t('forms.harbor.wwwUrl')}
              invalid={!!errors.wwwUrl}
              invalidText={errors.wwwUrl}
            />
          </Section>

          <Section>
            <Field
              as={FileUpload}
              label={t('forms.harbor.imageFile')}
              maxSize={500 * 1000}
              name="imageFile"
              onChange={(value: File) => {
                setFieldValue('imageFile', value);
              }}
              required
            />
          </Section>

          <Section>
            <Field
              as={FileUpload}
              label={t('forms.harbor.harborMaps')}
              name="harborMaps"
              multiple
              onChange={(value: File[]) => {
                setFieldValue('harborMaps', value);
              }}
              required
            />
          </Section>

          <div className={styles.formActionButtons}>
            <Button color="supplementary" onClick={onCancel}>
              {t('common.cancel')}
            </Button>
            <Button type="submit" disabled={isSubmitting}>
              {t('common.save')}
            </Button>
          </div>
        </Form>
      )}
    </Formik>
  );
};

export default HarborEditForm;
