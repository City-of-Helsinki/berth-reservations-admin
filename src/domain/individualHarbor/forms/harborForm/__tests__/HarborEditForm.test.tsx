import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import waitForExpect from 'wait-for-expect';
import { MockedProvider } from '@apollo/react-testing';

import LoadingSpinner from '../../../../../common/spinner/LoadingSpinner';
import HarborEditForm from '../HarborEditForm';
import { HARBOR_FORM_QUERY } from '../queries';

const queryMock = {
  request: { query: HARBOR_FORM_QUERY, variables: { id: 'a' } },
  result: {
    data: {
      harbor: {
        __typename: 'HarborNode',
        id: 'a',
        properties: {
          __typename: 'HarborProperties',
          name: 'Test harbor',
          streetAddress: 'Test Address 1',
          zipCode: '00100',
          municipality: 'Helsinki',
          wwwUrl: 'https://hel.fi',
          imageFile: 'https://hel.fi',
          maps: [
            {
              __typename: 'HarborMapType',
              id: 'aaa',
              url: 'https://hel.ninja',
            },
          ],
        },
      },
    },
  },
};

describe('HarborEditForm', () => {
  const waitForContent = async (wrapper: ReactWrapper) => {
    await act(async () => {
      await waitForExpect(() => {
        wrapper.update();
        expect(wrapper.contains(<LoadingSpinner isLoading={true} />)).toBeFalsy();
      });
    });
  };

  it('initially renders loading spinner', () => {
    const wrapper = mount(
      <MockedProvider mocks={[queryMock]}>
        <HarborEditForm onCancel={jest.fn()} onSubmit={jest.fn()} harborId={'a'} />
      </MockedProvider>
    );
    expect(wrapper.contains(<LoadingSpinner isLoading={true} />)).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders content after loading', async () => {
    const wrapper = mount(
      <MockedProvider mocks={[queryMock]}>
        <HarborEditForm onCancel={jest.fn()} onSubmit={jest.fn()} harborId={'a'} />
      </MockedProvider>
    );
    await waitForContent(wrapper);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
