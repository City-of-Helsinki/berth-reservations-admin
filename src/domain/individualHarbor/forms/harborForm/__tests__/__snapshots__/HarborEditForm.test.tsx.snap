// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HarborEditForm initially renders loading spinner 1`] = `"<div class=\\"spinnerWrapper\\"><div class=\\"spinner\\"></div></div>"`;

exports[`HarborEditForm renders content after loading 1`] = `"<form action=\\"#\\" class=\\"harborEditForm\\"><h4 class=\\"text brand h4\\">SATAMA</h4><article class=\\"section\\"><section class=\\"body\\"><div class=\\"TextInput-module_root__1ceal text-input_hds-text-input__24bl6\\"><label for=\\"name\\" class=\\"TextInput-module_label__1NkZD text-input_hds-text-input__label__1-UoQ \\">Nimi</label><div class=\\"TextInput-module_inputWrapper__1ru3h text-input_hds-text-input__input-wrapper__2p7Uy\\"><input class=\\"TextInput-module_input__35GDy text-input_hds-text-input__input__2XAyI\\" id=\\"name\\" type=\\"text\\" name=\\"name\\" required=\\"\\" value=\\"Pajalahden venesatama (Meripuistotie) / Venesatama\\"></div></div></section></article><article class=\\"section\\"><section class=\\"body\\"><div class=\\"TextInput-module_root__1ceal text-input_hds-text-input__24bl6\\"><label for=\\"streetAddress\\" class=\\"TextInput-module_label__1NkZD text-input_hds-text-input__label__1-UoQ \\">Osoite</label><div class=\\"TextInput-module_inputWrapper__1ru3h text-input_hds-text-input__input-wrapper__2p7Uy\\"><input class=\\"TextInput-module_input__35GDy text-input_hds-text-input__input__2XAyI\\" id=\\"streetAddress\\" type=\\"text\\" name=\\"streetAddress\\" required=\\"\\" value=\\"Meripuistotie 1a\\"></div></div><div class=\\"TextInput-module_root__1ceal text-input_hds-text-input__24bl6\\"><label for=\\"zipCode\\" class=\\"TextInput-module_label__1NkZD text-input_hds-text-input__label__1-UoQ \\">Postinumero</label><div class=\\"TextInput-module_inputWrapper__1ru3h text-input_hds-text-input__input-wrapper__2p7Uy\\"><input class=\\"TextInput-module_input__35GDy text-input_hds-text-input__input__2XAyI\\" id=\\"zipCode\\" type=\\"text\\" name=\\"zipCode\\" required=\\"\\" value=\\"00210\\"></div></div><div class=\\"TextInput-module_root__1ceal text-input_hds-text-input__24bl6\\"><label for=\\"municipality\\" class=\\"TextInput-module_label__1NkZD text-input_hds-text-input__label__1-UoQ \\">Postitoimipaikka</label><div class=\\"TextInput-module_inputWrapper__1ru3h text-input_hds-text-input__input-wrapper__2p7Uy\\"><input class=\\"TextInput-module_input__35GDy text-input_hds-text-input__input__2XAyI\\" id=\\"municipality\\" type=\\"text\\" name=\\"municipality\\" required=\\"\\" value=\\"Helsinki\\"></div></div></section></article><article class=\\"section\\"><section class=\\"body\\"><div class=\\"TextInput-module_root__1ceal text-input_hds-text-input__24bl6\\"><label for=\\"wwwUrl\\" class=\\"TextInput-module_label__1NkZD text-input_hds-text-input__label__1-UoQ \\">WWW-osoite</label><div class=\\"TextInput-module_inputWrapper__1ru3h text-input_hds-text-input__input-wrapper__2p7Uy\\"><input class=\\"TextInput-module_input__35GDy text-input_hds-text-input__input__2XAyI\\" id=\\"wwwUrl\\" type=\\"text\\" name=\\"wwwUrl\\" required=\\"\\" value=\\"https://hel.fi\\"></div></div></section></article><article class=\\"section\\"><section class=\\"body\\"><div class=\\"fileUpload\\"><span class=\\"labelText\\">Sataman kuva</span><ul class=\\"list standard noBullets\\"><li class=\\"fileListItem\\"><span class=\\"text brand span fileName\\">https://hel.fi</span><button class=\\"delete\\" type=\\"button\\"><div class=\\"icon critical medium\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 480 480\\" class=\\"\\"><path d=\\"M350.66 70.31a123.85 123.85 0 0 1-83-31.75L256 28.08l-11.65 10.48a123.85 123.85 0 0 1-83 31.75H93.44l3.07 34.84h64.84A158.75 158.75 0 0 0 256 74a158.75 158.75 0 0 0 94.66 31.17h64.83l3.07-34.84zM128.28 465.13h255.45L412.36 140H99.59zm246.05-290.29l-22.5 255.45H311l7.35-220.62h-34.89l-7.34 220.62h-40.24l-7.34-220.62h-34.86L201 430.29h-40.83l-22.54-255.45z\\"></path></svg></div></button></li></ul><div class=\\"row\\"><label class=\\"field\\"><input accept=\\"image/png, image/jpeg\\" class=\\"input\\" type=\\"file\\" name=\\"imageFile\\"><div class=\\"button supplementary small\\"><span class=\\"label\\">Valitse...</span></div></label><span class=\\"text secondary span maxSize\\">Tiedostokoko enint채채n 500 kB</span></div></div></section></article><article class=\\"section\\"><section class=\\"body\\"><div class=\\"fileUpload\\"><span class=\\"labelText\\">Satamakartat</span><ul class=\\"list standard noBullets\\"><li class=\\"fileListItem\\"><span class=\\"text brand span fileName\\">https://hel.ninja</span><button class=\\"delete\\" type=\\"button\\"><div class=\\"icon critical medium\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 480 480\\" class=\\"\\"><path d=\\"M350.66 70.31a123.85 123.85 0 0 1-83-31.75L256 28.08l-11.65 10.48a123.85 123.85 0 0 1-83 31.75H93.44l3.07 34.84h64.84A158.75 158.75 0 0 0 256 74a158.75 158.75 0 0 0 94.66 31.17h64.83l3.07-34.84zM128.28 465.13h255.45L412.36 140H99.59zm246.05-290.29l-22.5 255.45H311l7.35-220.62h-34.89l-7.34 220.62h-40.24l-7.34-220.62h-34.86L201 430.29h-40.83l-22.54-255.45z\\"></path></svg></div></button></li></ul><div class=\\"row\\"><label class=\\"field\\"><input class=\\"input\\" type=\\"file\\" name=\\"maps\\" multiple=\\"\\"><div class=\\"button supplementary small\\"><span class=\\"label\\">Lis채채...</span></div></label></div></div></section></article><div class=\\"formActionButtons\\"><button type=\\"button\\" class=\\"Button-module_button__3wgee button_hds-button__2A0je Button-module_secondary__399Vd button_hds-button--secondary__1NOWS Button-module_theme-black__zO4WO button_hds-button--theme-black__3K_k4\\"><span class=\\"Button-module_label__2zMLC button_hds-button__label__2EQa-\\">Peruuta</span></button><button type=\\"submit\\" class=\\"Button-module_button__3wgee button_hds-button__2A0je Button-module_primary__3bo_T button_hds-button--primary__2NVvO Button-module_theme-coat__2xA5D button_hds-button--theme-coat__12cO0\\"><span class=\\"Button-module_label__2zMLC button_hds-button__label__2EQa-\\">Tallenna</span></button></div></form>"`;
